archive/overlap.go

```diff
@@ -0,0 +1,135 @@
+package archive
+
+import (
+	"sort"
+
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/zbuf"
+)
+
+// mergeChunksToSpans takes an unordered set of Chunks with possibly overlapping
+// spans, and returns an ordered list of SpanInfos, whose spans will be bounded
+// by filter, and where each spanInfo contains one or more Chunks whose data
+// falls into the spanInfo's span.
+func mergeChunksToSpans(chunks []Chunk, dir zbuf.Direction, filter nano.Span) []spanInfo {
+	var siChunks []Chunk // accumulating chunks for next spanInfo
+	var siFirst nano.Ts  // first timestamp for next spanInfo
+	var result []spanInfo
+	boundaries(chunks, dir, func(ts nano.Ts, firstChunks, lastChunks []Chunk) {
+		if len(firstChunks) > 0 {
+			// ts is the 'First' timestamp for these chunks.
+			if len(siChunks) > 0 {
+				// We have accumulated chunks; create a span with them whose
+				// last timestamp was just before ts.
+				siSpan := closedSpan(siFirst, prevTs(ts, dir))
+				if filter.Overlaps(siSpan) {
+					result = append(result, spanInfo{
+						span:   filter.Intersect(siSpan),
+						chunks: copyChunks(siChunks, nil),
+					})
+				}
+			}
+			// Accumulate these chunks whose first timestamp is ts.
+			siChunks = append(siChunks, firstChunks...)
+			siFirst = ts
+		}
+		if len(lastChunks) > 0 {
+			// ts is the 'Last' timestamp for these chunks.
+			siSpan := closedSpan(siFirst, ts)
+			if filter.Overlaps(siSpan) {
+				result = append(result, spanInfo{
+					span:   filter.Intersect(siSpan),
+					chunks: copyChunks(siChunks, nil),
+				})
+			}
+			// Drop the chunks that ended from our accumulation.
+			siChunks = copyChunks(siChunks, lastChunks)
+			siFirst = nextTs(ts, dir)
+		}
+	})
+	return result
+}
+
+func copyChunks(src []Chunk, skip []Chunk) (dst []Chunk) {
+outer:
+	for i := range src {
+		for j := range skip {
+			if src[i].Id == skip[j].Id {
+				continue outer
+			}
+		}
+		dst = append(dst, src[i])
+	}
+	return
+}
+
+// closedSpan returns a span for the closed interval of [x,y].
+func closedSpan(x, y nano.Ts) nano.Span {
```

Comment:
Span aren't closed or open—they're always half-open—so this doesn't seem like the best name for this. How about firstLastToSpan for symmetry with spanToFirstLast?